## 연속형 quest

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
mpl.rc('font',family='Malgun Gothic')
import seaborn as sns
import scipy 
scipy.__version__
import scipy.stats as stats

In [2]:
df_TOCC = pd.read_csv('../../datasets/TypeOfContractChannel.csv')
df_TOCC[:3]

id type_of_contract type_of_contract2 channel    datetime  Term  \
0  66758234               렌탈            Normal  서비스 방문  2019-10-20    60   
1  66755948               렌탈  Extension_Rental  서비스 방문  2019-10-20    60   
2  66756657               렌탈            Normal  홈쇼핑/방송  2019-10-20    60   

  payment_type product  amount state  overdue_count overdue  credit rating  \
0          CMS      K1   96900  계약확정              0      없음            9.0   
1         카드이체      K1  102900  계약확정              0      없음            2.0   
2          CMS      K1   96900  계약확정              0      없음            8.0   

    bank cancellation   age  Mileage  
0  새마을금고           정상  43.0   1862.0  
1   현대카드           정상  62.0   2532.0  
2   우리은행           정상  60.0   2363.0

In [3]:
df_TOCC['credit rating'].value_counts()

1.0     10778
2.0      8228
4.0      5379
3.0      4997
5.0      4317
6.0      3818
7.0      2717
8.0      1542
9.0       639
0.0        60
10.0       46
Name: credit rating, dtype: int64

#### amount, Mileage 관계 p-value

#### 선정변수 
        1. 'amount' (연속), 'Mileage' (연속)
            → 분석 내용: amount에 따른 mileage의 변화?
                - 귀무가설: amount에 따른 mileage의 변화 없음
                - 대립가설: amount에 따른 mileage의 변화 있음
            → 분석 결론 (corr : 1? -1?, p-value : 0.05 ?)
            Results: PearsonRResult(statistic=-0.004447823574819926, pvalue=0.37086033933462437)
                - p-value는 0.37이고 corr는 -0.0044
                - 통계 결론: 귀무가설 참. 
                - 사실 결론: amount에 따른 mileage 변화는 없다.
                



In [4]:
df_TOCC.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51304 entries, 0 to 51303
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 51304 non-null  int64  
 1   type_of_contract   51300 non-null  object 
 2   type_of_contract2  51303 non-null  object 
 3   channel            51304 non-null  object 
 4   datetime           51304 non-null  object 
 5   Term               51304 non-null  int64  
 6   payment_type       51304 non-null  object 
 7   product            51303 non-null  object 
 8   amount             51304 non-null  int64  
 9   state              51304 non-null  object 
 10  overdue_count      51304 non-null  int64  
 11  overdue            51302 non-null  object 
 12  credit rating      42521 non-null  float64
 13  bank               48544 non-null  object 
 14  cancellation       51279 non-null  object 
 15  age                40509 non-null  float64
 16  Mileage            405

In [5]:
df_TOCC_drop=df_TOCC.dropna().copy()
df_TOCC_drop

id type_of_contract type_of_contract2 channel    datetime  Term  \
0      66758234               렌탈            Normal  서비스 방문  2019-10-20    60   
1      66755948               렌탈  Extension_Rental  서비스 방문  2019-10-20    60   
2      66756657               렌탈            Normal  홈쇼핑/방송  2019-10-20    60   
3      66423450              멤버십               TAS   렌탈재계약  2019-10-20    12   
4      66423204              멤버십               TAS   렌탈재계약  2019-10-20    12   
...         ...              ...               ...     ...         ...   ...   
51295  66799197               렌탈         Promotion    영업방판  2020-02-03    39   
51296  66792778               렌탈            Normal  홈쇼핑/방송  2020-02-03    60   
51297  66799607               렌탈            Normal  홈쇼핑/방송  2020-02-03    60   
51300  66799558               렌탈            Normal    영업방판  2020-02-03    60   
51303  66799607               렌탈            Normal    하이마트  2020-02-03    60   

      payment_type product  amount state  overdue_count overdue  \
0              CMS      K1   96900  계약확정              0      없음   
1             카드이체      K1  102900  계약확정              0      없음   
2              CMS      K1   96900  계약확정              0      없음   
3              CMS      K1   66900  계약확정              0      없음   
4              CMS      K1   66900  해약확정             12      있음   
...            ...     ...     ...   ...            ...     ...   
51295         카드이체      K5  120900  해약확정              0      없음   
51296         카드이체      K1   96900  계약확정              0      없음   
51297          CMS      K1   96900  해약확정              0      없음   
51300         카드이체      K1   96900  해약확정              0      없음   
51303          CMS      K1   96900  해약확정              0      없음   

       credit rating   bank cancellation   age  Mileage  
0                9.0  새마을금고           정상  43.0   1862.0  
1                2.0   현대카드           정상  62.0   2532.0  
2                8.0   우리은행           정상  60.0   2363.0  
3                5.0   농협은행           정상  60.0   2449.0  
4                8.0   농협은행           해약  51.0   1942.0  
...              ...    ...          ...   ...      ...  
51295            1.0   롯데카드           해약  48.0   1869.0  
51296            2.0   롯데카드           정상  44.0   1947.0  
51297            8.0   신한은행           해약  70.0   2623.0  
51300            8.0   롯데카드           해약  39.0   1753.0  
51303            8.0   농협은행           해약  53.0   2130.0  

[40480 rows x 17 columns]

In [6]:
df_TOCC_amount_Mileage = df_TOCC_drop[['amount', 'Mileage']]
df_TOCC_amount_Mileage

amount  Mileage
0       96900   1862.0
1      102900   2532.0
2       96900   2363.0
3       66900   2449.0
4       66900   1942.0
...       ...      ...
51295  120900   1869.0
51296   96900   1947.0
51297   96900   2623.0
51300   96900   1753.0
51303   96900   2130.0

[40480 rows x 2 columns]

In [7]:
df_TOCC_amount_Mileage.corr()

amount   Mileage
amount   1.000000 -0.004448
Mileage -0.004448  1.000000

In [8]:
stats.pearsonr(df_TOCC_amount_Mileage['amount'],df_TOCC_amount_Mileage['Mileage'])

PearsonRResult(statistic=-0.004447823574819926, pvalue=0.37086033933462437)

In [37]:
df_TOCC.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51304 entries, 0 to 51303
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 51304 non-null  int64  
 1   type_of_contract   51300 non-null  object 
 2   type_of_contract2  51303 non-null  object 
 3   channel            51304 non-null  object 
 4   datetime           51304 non-null  object 
 5   Term               51304 non-null  int64  
 6   payment_type       51304 non-null  object 
 7   product            51303 non-null  object 
 8   amount             51304 non-null  int64  
 9   state              51304 non-null  object 
 10  overdue_count      51304 non-null  int64  
 11  overdue            51302 non-null  object 
 12  credit rating      42521 non-null  float64
 13  bank               48544 non-null  object 
 14  cancellation       51279 non-null  object 
 15  age                40509 non-null  float64
 16  Mileage            405

In [40]:
df_TOCC_aM = df_TOCC_drop[['amount', 'Mileage']]
df_TOCC_aM[:4]

amount  Mileage
0   96900   1862.0
1  102900   2532.0
2   96900   2363.0
3   66900   2449.0

In [41]:
df_TOCC_aM.corr()

amount   Mileage
amount   1.000000 -0.004448
Mileage -0.004448  1.000000

In [42]:
stats.pearsonr(df_TOCC_aM['amount'],df_TOCC_aM['Mileage'])

PearsonRResult(statistic=-0.004447823574819926, pvalue=0.37086033933462437)

#### 선정변수 
        1. 'age' (연속), 'Mileage' (연속)
            → 분석 내용: 나이에 따른 마일리지 변화?
                - 귀무가설: 나이에 따른 마일리지 변화 없음.
                - 대립가설: 나이에 따른 마일리지 변화 있음.
            → 분석 결론 (p-value : 0.05 ?), stats.pearsonr('비교1', '비교2')
                ⇒ PearsonRResult(statistic=0.9732809346109912, pvalue=0.0)
                - 통계 결론: statistic_0.97, p-value < 0.05 이므로 대립가설 참
                - 사실 결론: 나이에 따른 마일리지 변화 있다.
            → correlation의 유의미함을 보려면 pearsonr을 사용한다.   

In [9]:
df_TOCC[['age','Mileage']].describe()
#count, mean, 50% 확인하면됨

age       Mileage
count  40509.000000  40509.000000
mean      46.608285   1967.922412
std       13.465977    436.707329
min       20.000000    872.000000
25%       37.000000   1646.000000
50%       46.000000   1946.000000
75%       56.000000   2263.000000
max      107.000000   3803.000000

In [10]:
df_TOCC[['age']].count()

age    40509
dtype: int64

In [11]:
df_TOCC[['Mileage']].count()

Mileage    40509
dtype: int64

In [12]:
df_age_Mileage = df_TOCC[['age','Mileage']]
df_age_Mileage[:2]

age  Mileage
0  43.0   1862.0
1  62.0   2532.0

In [13]:
df_TOCC[['age','Mileage']].corr()

age   Mileage
age      1.000000  0.973281
Mileage  0.973281  1.000000

In [14]:
df_age_Mileage.isnull().sum()

age        10795
Mileage    10795
dtype: int64

In [15]:
df_age_Mileage = df_age_Mileage.dropna().copy()

#### corr() 확인했을때, age-Mileage:0.973281 
    → 분포적으로 유의미한건가?, 어떤 가설이 맞는가?
    ⇒ correlation의 p-value를 보고자 확인: pearsonr 확인. 

In [16]:
df_age_Mileage.isnull().sum()

age        0
Mileage    0
dtype: int64

In [17]:
df_age_Mileage.corr()

age   Mileage
age      1.000000  0.973281
Mileage  0.973281  1.000000

In [18]:
stats.pearsonr(df_age_Mileage['age'], df_age_Mileage['Mileage'])

PearsonRResult(statistic=0.9732809346109912, pvalue=0.0)

    -> pearson에서 statistic = 0.97.. 이거의 음과 양의 수로 비례인지 반비례인지 확인 가능
    -> p-value= 0.0 으로 보아 유의미함

- 목표변수 먼저 설정하고 그때그때 가져오기

#### 선정변수 
        1. 'type_of_contract' (범주), 'cancellation' (범주)
            → 분석 내용: 계약 유형에 따른 해약 여부???
                - 귀무가설: 계약유형에 따른 해약여부는 달라지지 않음. 
                - 대립가설: 계약유형에 따른 해약여부는 달라짐. 
            → 분석 결론 (p-value : 0.05 ?)
                - 통계 결론: 대립가설 참
                - 사실 결론: 계약유형에 따라서 해약여부가 달라짐. 
                

In [19]:
df_TOCC[['type_of_contract','cancellation']].describe()
# 항목 개수 확인

type_of_contract cancellation
count             51300        51279
unique                2            2
top                  렌탈           정상
freq              46481        50372

In [20]:
df_TOCC_category = df_TOCC[['type_of_contract','cancellation']]
df_TOCC_category[:2]

type_of_contract cancellation
0               렌탈           정상
1               렌탈           정상

In [21]:
# index = groupby= y = 목표변수
df_TOCC_category_pivot = pd.pivot_table(data=df_TOCC_category, index=['type_of_contract'], columns=['cancellation'], aggfunc=(len))
df_TOCC_category_pivot

cancellation         정상   해약
type_of_contract            
렌탈                46137  344
멤버십                4231  563

#### pivot_table로는 숫자만 보여서 검증이 필요함. 
        - chi^2 을 사용

In [22]:
import scipy 
scipy.__version__
import scipy.stats as stats

- chi2_contingency(observed[, correction, lambda_])
- observedarray_like
The contingency table. The table contains the observed frequencies (i.e. number of occurrences) in each category. In the two-dimensional case, the table is often described as an “R x C table”.

In [23]:
stats.chi2_contingency(observed=df_TOCC_category_pivot, correction=True)

Chi2ContingencyResult(statistic=3021.9735309482894, pvalue=0.0, dof=1, expected_freq=array([[45658.8007411,   822.1992589],
       [ 4709.1992589,    84.8007411]]))

#### 카이제곱 결과 
        - p-value < 0.05 이므로 대립가설 참. 
        - 계약 유형에 따라 해약여부가 달라짐. 

In [24]:
df_TOCC[['state','cancellation']].describe()

state cancellation
count   51304        51279
unique      4            2
top      계약확정           정상
freq    50350        50372

#### 선정변수 
        1. 'product' (범주), 'cancellation' (범주)
            → 분석 내용: 차종에 따른 해약 여부 달라지는가?
                - 귀무가설: 차종에 따른 해약여부는 달라지지 않음. 
                - 대립가설: 차종에 따른 해약여부는 달라짐. 
            → 분석 결론 (p-value : 0.05 ?)
                - 통계 결론: 대립 가설 참.
                - 사실 결론: 차종에 따른 해약여부가 달라진다. 
                

In [25]:
df_TOCC_product=df_TOCC[['product','cancellation']].copy()
df_TOCC_product[:2]
# .copy(): 다른 변수 이름 작성 시 , 사용. 같은 영역 공유?
df_TOCC_product.describe()

product cancellation
count    51303        51279
unique       6            2
top         K1           정상
freq     39134        50372

In [26]:
# info로 null 있는지 확인
df_TOCC[['product','cancellation']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51304 entries, 0 to 51303
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product       51303 non-null  object
 1   cancellation  51279 non-null  object
dtypes: object(2)
memory usage: 801.8+ KB


In [27]:
# null 이 있는 것을 dropna()시켜버림. 
df_TOCC_product = df_TOCC[['product','cancellation']].dropna().copy()
df_TOCC_product

product cancellation
0          K1           정상
1          K1           정상
2          K1           정상
3          K1           정상
4          K1           해약
...       ...          ...
51299      K3           정상
51300      K1           해약
51301      K5           해약
51302      K1           정상
51303      K1           해약

[51278 rows x 2 columns]

In [28]:
# null값없는지 re-check
df_TOCC_product.isnull().sum()

product         0
cancellation    0
dtype: int64

In [29]:
df_TOCC_product_pivot= pd.pivot_table(data=df_TOCC_product, index='product',columns='cancellation', aggfunc=len)
df_TOCC_product_pivot

cancellation     정상   해약
product                 
K1            38291  820
K2             8926   67
K3             2072   10
K4              323    4
K5              641    4
K6              118    2

In [30]:
stats.chi2_contingency(observed=df_TOCC_product_pivot, correction=True,)

Chi2ContingencyResult(statistic=103.79474317337241, pvalue=8.372028115481391e-21, dof=5, expected_freq=array([[3.84192086e+04, 6.91791353e+02],
       [8.83393274e+03, 1.59067261e+02],
       [2.04517380e+03, 3.68262023e+01],
       [3.21216058e+02, 5.78394243e+00],
       [6.33591306e+02, 1.14086938e+01],
       [1.17877452e+02, 2.12254768e+00]]))

#### 선정변수 (chi2이 적용되지 않는경우)
        1. 'state' (범주), 'cancellation' (범주)
            → 분석 내용: 햬약상태에 따른 해약 여부 달라지는가?
                - chi2이 적용되지 않음. 
### 결론: 두 변수 구성은 참조되지 않음.                 

In [31]:
df_TOCC_error = df_TOCC[['state','cancellation']]
# df_TOCC_error[:3]
df_TOCC_error.describe()

state cancellation
count   51304        51279
unique      4            2
top      계약확정           정상
freq    50350        50372

In [32]:
df_TOCC_error = df_TOCC[['state','cancellation']].dropna().copy()
df_TOCC_error.describe()

state cancellation
count   51279        51279
unique      4            2
top      계약확정           정상
freq    50326        50372

In [33]:
df_TOCC_error_pivot=pd.pivot_table(data=df_TOCC_error, index='state', columns='cancellation', aggfunc=len)
df_TOCC_error_pivot

cancellation       정상     해약
state                       
계약확정          50326.0    NaN
기간만료             32.0    NaN
해약진행중            14.0    NaN
해약확정              NaN  907.0

In [34]:
stats.chi2_contingency(observed=df_TOCC_error_pivot, correction=True)

Chi2ContingencyResult(statistic=nan, pvalue=nan, dof=3, expected_freq=array([[nan, nan],
       [nan, nan],
       [nan, nan],
       [nan, nan]]))

chi2이 적용되지 않음. 

In [35]:
df_TOCC_errors_pivot=pd.pivot_table(data=df_TOCC_error, index='cancellation', columns='state', aggfunc=len)
df_TOCC_errors_pivot

state            계약확정  기간만료  해약진행중   해약확정
cancellation                             
정상            50326.0  32.0   14.0    NaN
해약                NaN   NaN    NaN  907.0

In [36]:
stats.chi2_contingency(observed=df_TOCC_errors_pivot, correction=True)

Chi2ContingencyResult(statistic=nan, pvalue=nan, dof=3, expected_freq=array([[nan, nan, nan, nan],
       [nan, nan, nan, nan]]))

chi2 확인 불가임.